syncd/syncd_flex_counter.cpp

```diff
@@ -597,6 +645,44 @@ void FlexCounter::removePort(
     }
 }
 
+void FlexCounter::removePortDebugCounters(
+        _In_ sai_object_id_t portVid,
+        _In_ std::string instanceId)
+{
+    SWSS_LOG_ENTER();
+
+    FlexCounter &fc = getInstance(instanceId);
+
+    std::unique_lock<std::mutex> lkMgr(fc.m_mtx);
+
+    auto it = fc.m_portDebugCounterIdsMap.find(portVid);
+    if (it == fc.m_portDebugCounterIdsMap.end())
+    {
+        SWSS_LOG_NOTICE("Trying to remove nonexisting port debug counter Ids 0x%" PRIx64, portVid);
+
+        // Remove flex counter if all counter IDs and plugins are unregistered
+        if (fc.isEmpty())
+        {
+            lkMgr.unlock();
```

Comment:
i would suggest to avoid manual unlocking locks, since this leads to some complicated scenarios, when for example exception is throw, or someone else starts modifying the code and he don't know about whats protected by lock and where it should be released. Its better to create samaller method that will have scope lock on mutex protecting the only thing that needs to be protected. But this entire flex counters needs to refactored anyway, locks are used here in more than 20 places which is overkill

Reply:
I completely agree. I'll go ahead and refactor the new methods for debug counters to follow this convention, and make a note to go back and clean-up the lock usage at some point in the future.

Reply:
There's actually a much bigger problem here that I did not notice before - none of the collectCounters methods acquires the lock before iterating over the counter ID map. I'll probably raise a separate PR for this.

Reply:
So after taking a closer look at this issue, it seems that right now the manual locking/unlocking is necessary because of this line here: https://github.com/Azure/sonic-sairedis/blob/89b4dc013070b1951daf22d26b3ca025da3146c0/syncd/syncd_flex_counter.cpp#L1164

Because this method is returning the raw pointer instead of a shared pointer the call to removeInstance will delete the flex counter, which may cause issues if we're still holding the lock.

@kcudnik I think maybe we should continue to follow what the other flex counter methods are doing for this particular change and then proceed with a refactor to fix the locking and shared pointer issues in a later PR. What do you think?